### ISSUE TITLE
Fix data type for code_message in twisted.web.http

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7981]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)| @adiroiban reported|
|-|-|
|Trac ID|trac#7981|
|Type|defect|
|Created|2015-07-25 13:48:15Z|
|Branch|https://github.com/twisted/twisted/tree/http-codemessage-7981|

Based on [#7939](https://github.com/twisted/twisted/issues/7939)#comment:9

-----------

docstring inform that code_message should be bytes but _responses.py and setResponseCode method handles it as text/Unicode.

---------

twisted/web/http.py#L967 is a bug and should be fixed in a separate ticket.

Looks like we are missing a test.

I would prefer to keep code_message as text/unicode and only convert it to bytes before putting it to the wire.

This mean that setResponseCode docstring should be updated to inform that message is str/text but only ascii characters and networkString is left as it is.

I think that using text(ascii) in setResponseCode is a better API.
Forcing it to bytes we would need to fix all messages from _responses.py and then to encode to ascii each time we cal setResponseCode.

The message is supposed to be a human readable text, so I think that Unicode is the right data type... but we should document that only ascii characters are supported.

Attachments:

* [http-response-message.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/43e/43e94e04ee17bbb3bf844b701b5c11e11725adfb/7fc9bb639b3b4abcb60d96e140c632c5ae2aef9e.patch) (812 bytes) - added by _sunu_ on 2015-07-28 15:58:42Z - 
* [http-response-message-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/43e/43e94e04ee17bbb3bf844b701b5c11e11725adfb/500bcd94f5c3098c1bdcf19fc5277ef80c49c302.patch) (2637 bytes) - added by _sunu_ on 2015-08-01 17:54:15Z - 
* [http-response-message-3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/43e/43e94e04ee17bbb3bf844b701b5c11e11725adfb/1bd64f12ca2bc5e0e5e5d0b9209ec6f22d9484fe.patch) (6489 bytes) - added by _sunu_ on 2015-08-13 15:09:07Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7981 7981
type__defect defect
reporter__adiroiban adiroiban
priority__normal normal
milestone__ 
branch__branches_http_codemessage_7981 branches/http-codemessage-7981
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1437832095584018 1437832095584018
changetime__1439618034192063 1439618034192063
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="sunu's avatar" src="https://avatars.githubusercontent.com/u/1142203?s=50" width="50" height="50">](https://github.com/sunu)<a name="note_1"></a>|@sunu removed owner|
|-|-|

Apart from changing the docstring to mention that we expect message to be ascii text I am also converting any message of type byte to unicode because some parts like https://github.com/twisted/twisted/blob/e6df2c23df4aae37e9d54adb0785ac90a699653c/twisted/web/http.py#L446 for example, use a byte message while calling this function.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban set owner to @sunu|
|-|-|

Many thanks for you patch!

I don't like python code which makes use of isinstance and I don't like code which tries to recover from bad input :)

For backward compatibility we might need to support on the fly type conversion, but this should be done with a deprecation warning so that we don't encourage the usage with bad types.

The new 'if branch' need an associated test with a docstring describing why we need this behaviour.

If deprecation code is added, the deprecation will also need a test.

Why not fix the http.py#L466 call so that it will always call this method with str/unicode as documented?

I think that the bug is not here but in the caller code.

Please check my commend and add the required test :)

Many thanks for you contribution and feedback!

**Comment by twisted-trac** [comment]
|[<img alt="sunu's avatar" src="https://avatars.githubusercontent.com/u/1142203?s=50" width="50" height="50">](https://github.com/sunu)<a name="note_3"></a>|@sunu removed owner|
|-|-|

Hey adiroiban!

I added a deprecation warning and a test. Now the problem is, the tests for response message fail on Python2 because on Python 2 networkString expects a (byte)string and returns unicode (where as on Python 2 it expects unicode and returns bytes). What's a reasonable way to deal with that?

Second problem is that if I fix the call at https://github.com/twisted/twisted/blob/e6df2c23df4aae37e9d54adb0785ac90a699653c/twisted/web/http.py#L446 (which I have done in this patch), some tests like https://github.com/twisted/twisted/blob/e6df2c23df4aae37e9d54adb0785ac90a699653c/twisted/web/test/test_proxy.py#L293 fails on Python 3 because they expect a bytestring as status message. Should I fix those tests in this ticket?

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban set owner to @sunu|
|-|-|

If the code should not handle bytes, then the tests should be fixed.

`test_setResponseCodeByteMessage` should be named to `test_setResponseCodeByteMessage_bytes_deprecation`. In this way we can add more unit tests to setResponseCodeByteMessage without classing names.
Also the docstring of this test should state that bytes are accepted, warned and auto-converted into str.

You don't need `lambda: req.setResponseCode(202, b"happily accepted")`, just use `setResponseCode, 202, b"happily accepted"` in assertWarns.
You can also just call setResponseCode() and then get the warnings using flushWarnings.

Please check other codes raising deprecatiosn wanings. THe warning should include the version number since when this is deprecated.

Thanks for working on this.
I think that this going in the right direction.

I have also asked hawkowl to take a look at this as she knows more about twisted.web.

Please check my commend and come back with a new patch.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="sunu's avatar" src="https://avatars.githubusercontent.com/u/1142203?s=50" width="50" height="50">](https://github.com/sunu)<a name="note_5"></a>|@sunu removed owner|
|-|-|

As per our discussion on IRC, I have converted all the response messages to bytes.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl commented|
|-|-|

(In [45475]) Branching to http-codemessage-7981.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl set owner to @hawkowl|
|-|-|

This looks good to me. It's missing a topfile (which I added) and the "happily accepted" in a test where it's not didn't match, so I fixed that as well.

Thanks, will merge.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl set status to `closed`|
|-|-|

(In [45479]) Merge http-codemessage-7981: Make web.http.Request.setResponseCode() enforce bytes and have _responses return bytes

Author: sunu
Reviewers: adiroiban, hawkowl
Fixes: [#7981](https://github.com/twisted/twisted/issues/7981)
